src/package-resolver.js

```diff
@@ -530,10 +530,48 @@ export default class PackageResolver {
       this.resolveToResolution(req);
     }
 
+    deps.forEach(dep => {
+      const name = normalizePattern(dep.pattern).name;
+      this.optimizeResolutions(name);
+    });
+
     activity.end();
     this.activity = null;
   }
 
+  // for a given package, see if a single manifest can satisfy all ranges
+  optimizeResolutions(name: string) {
+    const patterns: Array<string> = this.patternsByPackage[name] || [];
+
+    // don't optimize things that already have a lockfile entry:
+    // https://github.com/yarnpkg/yarn/issues/79
+    const anyLocked = patterns.some(pattern => this.lockfile.getLocked(pattern));
+    if (anyLocked) {
+      return;
+    }
+
+    const manifestByVersion: Map<string, Manifest> = new Map();
+    patterns.forEach(p => {
+      const manifest = this.patterns[p];
+      manifestByVersion.set(manifest.version, manifest);
+    });
+
+    // reverse sort, so we'll find the maximum satisfying version
+    const availableVersions = [...manifestByVersion.keys()];
+    availableVersions.sort(semver.rcompare);
+
+    const requiredRanges = patterns.map(p => normalizePattern(p).range);
```

Comment:
I think you can leverage some semver secrets as follows:
```js
const fullRange = patterns.map(pattern => normalizePattern(p).range).join(' ');
const oneRingToRuleThemAll = semver.maxSatisfying([...manifestByVersion.keys()], fullRange);
if (oneRingToRuleThemAll) {
    this.collapseAllVersionsOfPackage(name, oneRingToRuleThemAll);
}
```

This should be more efficient than what the code below does. Also, the code below continuously resolves to new versions as it goes on since `return` in a `forEach` loop doesn't terminate the loop at all.

Reply:
Yep, that's a nice optimization.